common/utilities/number/stabilized-value.h

```diff
@@ -0,0 +1,136 @@
+// The get will return one of 3 values, depend on the case:
+//      case 1 : the highest occurrences value in the history meet the require percent stability - return it
+//      case 2 : if not , if the last stable value is still exist in the history, return it.
+//      case 3 : if the last stable value is not in the history, return the highest occurrences value
+// Inputs: Template <value_type>
+//         required history size
+//         required stabilization percentage as a fraction , range [0-1]
+// The stabilized_value class is thread safe.
+template < typename T > class stabilized_value
+{
+public:
+    stabilized_value( size_t history_size )
+        : _history_size( history_size )
+        , _last_stable_value( 0 )
+        , _last_stable_value_set( false )
+        , _recalc_stable_val( false )
+    {
+    }
+
+    stabilized_value() = delete;
+    stabilized_value(const stabilized_value &) = delete;
+
+
+    void add( T val )
+    {
+        const std::lock_guard< std::mutex > lock( _mutex );
+        if( _values.empty() )
+        {
+            _values.push_back( val );
+            _last_stable_value = val;
+            _recalc_stable_val = false;
+            return;
+        }
+
+        _recalc_stable_val = true;
+
+        _values.push_back( val );
+        if( _values.size() > _history_size )
+            _values.pop_front();
+    }
+
+    T get( float stabilization_percent = 0.75f) const
+    {
+        if ((stabilization_percent <= 0.0f) || (stabilization_percent > 1.0f))
+            throw std::runtime_error("Illegal value for stabilize_percentage: "
+                + std::to_string(stabilization_percent));
+
+        std::lock_guard< std::mutex > lock( _mutex );
+
+        if( _values.empty() )
+        {
+            throw std::runtime_error( "history is empty; no stable value" );
+        }
+
+        if (_recalc_stable_val || (_stabilize_percentage != stabilization_percent))
+        {
+            _stabilize_percentage = stabilization_percent;
+            std::unordered_map< T, int > values_count_map;
+            std::pair< T, int > most_stable_value = { 0, 0 };
+            for( T val : _values )
+            {
+                auto current_val = ++values_count_map[val];
+
+                if( most_stable_value.second < current_val)
+                {
+                    most_stable_value.first = val;
+                    most_stable_value.second = current_val;
+                }
+            }
+
+            auto new_value_percentage
+                = most_stable_value.second / static_cast< float >( _values.size() );
+            
+            // if the highest occurrences value meet the stabilized percentage request return it.
+            // If not, if the last stable value exist in the history, return it , if not return the best we got. (highest occurrences value)
+            if (!_last_stable_value_set || (new_value_percentage >= _stabilize_percentage))
```

Comment:
There's never a case where `_last_stable_value_set` will be `false` here... I think it should be completely removed